kyuubi-main/src/main/scala/org/apache/kyuubi/client/KyuubiSyncThriftClient.scala

```diff
@@ -0,0 +1,201 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.client
+
+import java.util.concurrent.locks.ReentrantLock
+
+import scala.collection.JavaConverters._
+
+import org.apache.hive.service.rpc.thrift._
+import org.apache.thrift.protocol.TProtocol
+
+import org.apache.kyuubi.KyuubiSQLException
+import org.apache.kyuubi.operation.FetchOrientation
+import org.apache.kyuubi.operation.FetchOrientation.FetchOrientation
+import org.apache.kyuubi.util.ThriftUtils
+
+class KyuubiSyncThriftClient(protocol: TProtocol) extends TCLIService.Client(protocol) {
+
+  @volatile private var _remoteSessionHandle: TSessionHandle = _
+
+  private val lock = new ReentrantLock()
+
+  /**
+   * Lock every rpc call to send them sequentially
+   */
+  private def withLockAcquired[T](block: => T): T = {
+    try {
+      lock.lock()
+      if (!protocol.getTransport.isOpen) {
+        throw KyuubiSQLException.connectionDoesNotExist()
+      }
+      block
+    } finally lock.unlock()
```

Comment:
seems the explicit `ReentrantLock` can be simplified by `synchronized { }` here.